File: cs/src/day03.cs
public class Day03 : DayBase
{
    public Day03() : base(3) {}

    public override void Run()
    {
        var lines = LoadLines().ToList();
        var ll = lines[0].Length;
        int lc = lines.Count, half = lc / 2;
        var bitcounts = new int[ll];
        lines.ForEach(line =>
        {
            for (var i = 0; i < ll; i++)
            {
                if (line[i] == '1')
                    bitcounts[i] += 1;
            }
        });
        var gamma = Convert.ToInt32(new string(bitcounts.Select(c => c > half ? '1' : '0').ToArray()), 2);
        var epsilon = Convert.ToInt32(new string(bitcounts.Select(c => c < half ? '1' : '0').ToArray()), 2);

        // Part 2
        var ogr = Convert.ToInt32(FilterLines(lines, true), 2);
        var csr = Convert.ToInt32(FilterLines(lines, false), 2);

        Console.WriteLine($"Day {DayNumber}");
        Console.WriteLine($"  Part 1: {gamma * epsilon}");
        Console.WriteLine($"  Part 2: {ogr * csr}");
    }

    private string FilterLines(IEnumerable<string> lines, bool getMost)
    {
        var filtered = lines.ToList();
        var index = 0;
        while (filtered.Count > 1)
        {
            var onecount = filtered.Count(line => line[index] == '1');
            var mostch = filtered.Count - onecount > onecount ? '0' : '1';
            var leastch = mostch == '1' ? '0' : '1';
            var ch = getMost ? mostch : leastch;
            filtered = filtered.Where(line => line[index] == ch).ToList();

            index++;
        }
        return filtered[0];
    }
}

File: py/day03.py
import os

def filterLines(lines: list[str], idx: int, selectMost: bool) -> list[str]:
    if len(lines) == 1:
        return lines

    zc = sum(line[idx] == '0' for line in lines)
    mostch = '1' if len(lines) - zc >= zc else '0'
    leastch = '0' if mostch == '1' else '1'
    selectch = mostch if selectMost else leastch
    return [line for line in lines if line[idx] == selectch]

def filterInput(lines: list[str], selectMost: bool) -> str:
    filtered = lines.copy()
    idx = 0
    while len(filtered) > 1:
        filtered = filterLines(filtered, idx, selectMost)
        idx += 1
    return filtered[0]

#
# Main
#
with open(os.path.dirname(os.path.realpath(__file__)) + "/../input/day03-input", "r") as f:
    input = [line.strip() for line in f.readlines()]

bitlen, linecount = len(input[0]), len(input)
counts = [0] * bitlen
for line in input:
    for idx in range(bitlen):
        if line[idx] == '1':
            counts[idx] += 1

# Take the counts array, treat > half as '1' and < half as '0' and convert 'binary' string to int
half = linecount / 2
mostCommon = int("".join(['1' if c > half else '0' for c in counts]), 2)
leastCommon = int("".join(['0' if c > half else '1' for c in counts]), 2)

# Part 2
mosts = filterInput(input, True)
leasts = filterInput(input, False)

print("Day 3")
print("  Part 1:", mostCommon * leastCommon)
print("  Part 2:", int(mosts, 2) * int(leasts, 2))
